### PR TITLE
[POC] AMP for TPUs

### PR DESCRIPTION
**Pull request by cowanmeg**
POC for enabling AMP on TPUs with bfloat16. 

HLO dump of MNIST model from test/test_train_mp_mnist_amp.py
```
ENTRY %IrToHlo.149 (p0.1: f32[10], p1.2: f32[10,50], p2.4: f32[50], p3.5: f32[50,320], p4.7: f32[20], p5.8: f32[20], p6.9: f32[20], p7.10: f32[20], p8.11: f32[20], p9.12: f32[20,10,5,5], p10.13: f32[10], p11.14: f32[10], p12.15: f32[10], p13.16: f32[10], p14.17: f32[10], p15.18: f32[10,1,5,5], p16.19: f32[128,1,28,28]) -> (f32[128,10]) {
  ...
  %p16.19 = f32[128,1,28,28]{0,3,2,1} parameter(16)
  %p15.18 = f32[10,1,5,5]{0,3,2,1} parameter(15)
  %convolution.20 = f32[128,10,24,24]{3,2,1,0} convolution(f32[128,1,28,28]{0,3,2,1} %p16.19, f32[10,1,5,5]{0,3,2,1} %p15.18), window={size=5x5}, dim_labels=bf01_oi01->bf01
  %p14.17 = f32[10]{0} parameter(14)
  %broadcast.21 = f32[128,24,24,10]{3,2,1,0} broadcast(f32[10]{0} %p14.17), dimensions={3}
  %transpose.22 = f32[128,10,24,24]{1,3,2,0} transpose(f32[128,24,24,10]{3,2,1,0} %broadcast.21), dimensions={0,3,1,2}
  %add.23 = f32[128,10,24,24]{3,2,1,0} add(f32[128,10,24,24]{3,2,1,0} %convolution.20, f32[128,10,24,24]{1,3,2,0} %transpose.22)
 ...
  %broadcast.136 = f32[128,10]{1,0} broadcast(f32[128]{0} %reduce.135), dimensions={0}
  %subtract.137 = f32[128,10]{1,0} subtract(f32[128,10]{1,0} %add.129, f32[128,10]{1,0} %broadcast.136)
  %exponential.138 = f32[128,10]{1,0} exponential(f32[128,10]{1,0} %subtract.137)
  %constant.139 = f32[] constant(0)
  %reduce.144 = f32[128]{0} reduce(f32[128,10]{1,0} %exponential.138, f32[] %constant.139), dimensions={1}, to_apply=%AddComputation.140
  %log.145 = f32[128]{0} log(f32[128]{0} %reduce.144)
  %broadcast.146 = f32[128,10]{1,0} broadcast(f32[128]{0} %log.145), dimensions={0}
  %subtract.147 = f32[128,10]{1,0} subtract(f32[128,10]{1,0} %subtract.137, f32[128,10]{1,0} %broadcast.146)
  ROOT %tuple.148 = (f32[128,10]{1,0}) tuple(f32[128,10]{1,0} %subtract.147)
}

}
```

---

HLO dump with AMP
```
ENTRY %IrToHlo.162 (p0.1: f32[10], p1.3: f32[10,50], p2.6: f32[50], p3.8: f32[50,320], p4.11: f32[20], p5.12: f32[20], p6.13: f32[20], p7.14: f32[20], p8.15: f32[20], p9.17: f32[20,10,5,5], p10.19: f32[10], p11.20: f32[10], p12.21: f32[10], p13.22: f32[10], p14.23: f32[10], p15.25: f32[10,1,5,5], p16.27: f32[128,1,28,28]) -> (bf16[128,10]) {
 ...
  %p16.27 = f32[128,1,28,28]{0,3,2,1} parameter(16)
  %convert.28 = bf16[128,1,28,28]{0,3,2,1} convert(f32[128,1,28,28]{0,3,2,1} %p16.27)
  %p15.25 = f32[10,1,5,5]{0,3,2,1} parameter(15)
  %convert.26 = bf16[10,1,5,5]{0,3,2,1} convert(f32[10,1,5,5]{0,3,2,1} %p15.25)
  %convolution.29 = bf16[128,10,24,24]{3,2,1,0} convolution(bf16[128,1,28,28]{0,3,2,1} %convert.28, bf16[10,1,5,5]{0,3,2,1} %convert.26), window={size=5x5}, dim_labels=bf01_oi01->bf01
  %p14.23 = f32[10]{0} parameter(14)
  %convert.24 = bf16[10]{0} convert(f32[10]{0} %p14.23)
  %broadcast.30 = bf16[128,24,24,10]{3,2,1,0} broadcast(bf16[10]{0} %convert.24), dimensions={3}
  %transpose.31 = bf16[128,10,24,24]{1,3,2,0} transpose(bf16[128,24,24,10]{3,2,1,0} %broadcast.30), dimensions={0,3,1,2}
  %add.32 = bf16[128,10,24,24]{3,2,1,0} add(bf16[128,10,24,24]{3,2,1,0} %convolution.29, bf16[128,10,24,24]{1,3,2,0} %transpose.31)
  ...
  %broadcast.149 = bf16[128,10]{1,0} broadcast(bf16[128]{0} %reduce.148), dimensions={0}
  %subtract.150 = bf16[128,10]{1,0} subtract(bf16[128,10]{1,0} %add.142, bf16[128,10]{1,0} %broadcast.149)
  %exponential.151 = bf16[128,10]{1,0} exponential(bf16[128,10]{1,0} %subtract.150)
  %constant.152 = bf16[] constant(0)
  %reduce.157 = bf16[128]{0} reduce(bf16[128,10]{1,0} %exponential.151, bf16[] %constant.152), dimensions={1}, to_apply=%AddComputation.153
  %log.158 = bf16[128]{0} log(bf16[128]{0} %reduce.157)
  %broadcast.159 = bf16[128,10]{1,0} broadcast(bf16[128]{0} %log.158), dimensions={0}
  %subtract.160 = bf16[128,10]{1,0} subtract(bf16[128,10]{1,0} %subtract.150, bf16[128,10]{1,0} %broadcast.159)
  ROOT %tuple.161 = (bf16[128,10]{1,0}) tuple(bf16[128,10]{1,0} %subtract.160)
}

```

### CODE REVIEW

**Comment by cowanmeg** [comment]
@miladm @JackCaoG

**Comment by JackCaoG** [comment]
Do you have a corresponding upstream pr? I saw you use `torch.xla.amp.autocast` but I don't see where it is defined. Seems like it is part of the `torch`

File: test/test_train_mp_mnist_amp.py [hunk 1]
```diff
@@ -130,14 +130,21 @@ def train_mnist(flags, **kwargs):
...
   model = MNIST().to(device)
   writer = None
   if xm.is_master_ordinal():
     writer = test_utils.get_summary_writer(flags.logdir)
   optim_cls = syncfree.SGD if FLAGS.use_syncfree_optim else optim.SGD
   optimizer = optim_cls(model.parameters(), lr=lr, momentum=flags.momentum)
   loss_fn = nn.NLLLoss()
-  scaler = GradScaler(use_zero_grad=FLAGS.use_zero_grad)
+  if device_hw == 'TPU':
+        autocast = torch.xla.amp.autocast
```
Comment (miladm) [inline test/test_train_mp_mnist_amp.py:142]: Question: can you point me to the mapping the explains what op is `autocast`ed and what op is not?

**Comment by miladm** [comment]
@chandrasekhard2 @cowanmeg who is exploring the ResNet model test? is this part of this PR or a seperate PR?

**Comment by chandrasekhard2** [comment]
> @chandrasekhard2 @cowanmeg who is exploring the ResNet model test? is this part of this PR or a separate PR?

I will send a PR today. Thanks!

**Comment by miladm** [comment]
@chandrasekhard2 @cowanmeg can you please share a few words on how the AMP perf looks like at the moment? it's fine if we need more work to improve numbers.

**Comment by wonjoolee95** [comment]
I assume this PR needs your PyTorch changes -- you can pin the PyTorch PR in this PR so it builds with your PyTorch PR. [Example of PyTorch pinning here](https://github.com/pytorch/xla/pull/3792/commits/40f41fb98b0f2386d287eeac0bae86e873d4a9d8).

**Comment by miladm** [comment]
@cowanmeg can we do a operator by operator mapping of the current model vs. what we see in HLO to verify what we added to the upstream pytorch for AMP maps to what emerges on the HLO side?

File: test/test_train_mp_imagenet_amp.py [hunk 2]
```diff
@@ -219,7 +219,13 @@ def train_imagenet():
...
   loss_fn = nn.CrossEntropyLoss()
   if FLAGS.amp:
-    scaler = GradScaler(use_zero_grad=FLAGS.use_zero_grad)
+    if device_hw == 'TPU':
+        autocast = torch.xla.amp.autocast
+        scaler = None
+    elif device_hw == 'GPU':
+        autocast = torch.cuda.amp.autocast
+        # GradScaler only used for GPU
+        scaler = GradScaler(use_zero_grad=FLAGS.use_zero_grad)
```
Comment (steventk-g) [inline test/test_train_mp_imagenet_amp.py:228]: Why don't we use a scaler for TPU?

**Comment by chandrasekhard2** [comment]
> @chandrasekhard2 @cowanmeg who is exploring the ResNet model test? is this part of this PR or a separate PR?

https://github.com/pytorch/xla/pull/4776/files

File: test/test_train_mp_imagenet_amp.py (hunk 2 above)
Comment (cowanmeg) [reply test/test_train_mp_imagenet_amp.py:228]: That comment was a bit misleading. GradScaler is necessary for float16 loss but not bfloat16. float16 has less dynamic range than float32 since it uses only 5 exponent bits vs 8 exponent bits, so the loss needs to be scaled to prevent small values from disappearing. bfloat16 has the same dynamic range as float32 so doesn't have this problem.

File: test/test_train_mp_mnist_amp.py (hunk 1 above)
Comment (cowanmeg) [reply test/test_train_mp_mnist_amp.py:142]: The op list is registered on the pytorch side here: https://github.com/pytorch/pytorch/blob/8455aac198e4ee68003f2b38a5e15631fb82690c/aten/src/ATen/autocast_mode.cpp#L520

File: test/test_train_mp_imagenet_amp.py (hunk 2 above)
Comment (vanbasten23) [reply test/test_train_mp_imagenet_amp.py:228]: That means TPU uses bfloat16 by default?

File: torch_xla/csrc/autocast_mode.cpp [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                     // type-aware overload.
+  promote, // Run in the widest dtype among several args.
+};
+
+// Base template for WrapFunction_, which is specialized to contain a "call" method each CastPolicy
+template<CastPolicy policy, DeviceType device_type, class Redispatch, Redispatch* F, class Ret, class ArgList> struct WrapFunction_ {};
+
+// CastPolicy::lower_precision_fp General_DeviceType
+template<DeviceType device_type, class Redispatch, Redispatch* F, class Ret, class... Args>
+struct WrapFunction_<CastPolicy::lower_precision_fp, device_type, Redispatch, F, Ret, guts::typelist::typelist<Args...>> {
```
Comment (vanbasten23) [inline torch_xla/csrc/autocast_mode.cpp:40]: wondering how this file is being used in the PR

**Comment by steventk-g** [comment]
Could you add a few unit tests to the C++ or python side to exercise amp for a simple case, e.g. a single operation?

File: test/test_train_mp_imagenet_amp.py (hunk 2 above)
Comment (cowanmeg) [reply test/test_train_mp_imagenet_amp.py]: For AMP, yes!

File: torch_xla/csrc/autocast_mode.cpp (hunk 3 above)
Comment (cowanmeg) [reply torch_xla/csrc/autocast_mode.cpp:18]: These are templates that help dispatch an op to its lower precision/fp32/promoted version.

Comment (cowanmeg) [reply torch_xla/csrc/autocast_mode.cpp:18]: Is the proper place for this file?

File: test/run_tests.sh [hunk 4]
```diff
@@ -168,6 +168,7 @@ function run_op_tests {
   run_test "$CDIR/test_operations.py" "$@" --verbosity=$VERBOSITY
   run_test_without_functionalization "$CDIR/test_operations.py" "$@" --verbosity=$VERBOSITY
   run_test "$CDIR/test_async_closures.py"
+  run_test "$CDIR/test_autocast.py"
```
Comment (JackCaoG) [inline test/run_tests.sh:171]: Thanks, did you run this test on TPUVM to verified it passed as expected? I copied most of the test case from a gpu test, hence I am not sure if those test cases are also suitable for the TPU.

File: test/test_train_mp_imagenet_amp.py [hunk 5]
```diff
@@ -219,23 +219,29 @@ def train_imagenet():
       summary_writer=writer)
   loss_fn = nn.CrossEntropyLoss()
   if FLAGS.amp:
-    scaler = GradScaler(use_zero_grad=FLAGS.use_zero_grad)
+    if device_hw == 'TPU':
+      scaler = None
+    elif device_hw == 'GPU':
+      scaler = GradScaler(use_zero_grad=FLAGS.use_zero_grad)
```
Comment (JackCaoG) [inline test/test_train_mp_imagenet_amp.py:225]: Now we have a full AMP story, I think we should add a doc under https://github.com/pytorch/xla/tree/master/docs to explain how to use it on TPU and GPU. We should also mentioned tricks like `GradScaler` here. WDYT?

I had a https://github.com/pytorch/xla/blob/master/docs/gpu.md#amp-automatic-mixed-precision here but I think it is better to have a standalone doc we can refer to during release.

File: torch_xla/amp/autocast_mode.py [hunk 6]
```diff
@@ -1,5 +1,36 @@
...
+               enabled: bool = True,
+               dtype: torch.dtype = torch.bfloat16,
+               cache_enabled: bool = True):
+    if xm.xla_device_hw(device) == 'GPU':
+      super().__init__(
+          "cuda",
+          enabled=enabled,
+          dtype=torch.float16,
+          cache_enabled=cache_enabled)
+    else:
```
Comment (JackCaoG) [inline torch_xla/amp/autocast_mode.py:24]: hmm so CPU and TPU shared the same autocast rule here?

**Review by JackCaoG — COMMENTED** [review]
Mostly LGTM, this is exciting!

I have one ux problem, if user do `torch.amp.autocast('xla')` on GPU device, is there anyway for us to throw a warning and get them to either use `torch_xla`'s autocast or use cuda directly?

File: torch_xla/amp/autocast_mode.py (hunk 6 above)
Comment (cowanmeg) [reply torch_xla/amp/autocast_mode.py:24]: Good catch

File: test/test_train_mp_imagenet_amp.py (hunk 5 above)
Comment (cowanmeg) [reply test/test_train_mp_imagenet_amp.py:225]: I agree, will add some documentation.

**Comment by cowanmeg** [comment]
> Mostly LGTM, this is exciting!
> 
> I have one ux problem, if user do `torch.amp.autocast('xla')` on GPU device, is there anyway for us to throw a warning and get them to either use `torch_xla`'s autocast or use cuda directly?

That code will be in Pytorch, so I don't think we can generate a warning since there is no way distinguish XLA:GPU vs. XLA:TPU that I am aware of.

File: test/run_tests.sh (hunk 4 above)
Comment (cowanmeg) [reply test/run_tests.sh:171]: The TPU tests follow the same format as the GPU tests, but I modified the list of ops that are tested for TPUs. I did verify the TPU tests pass on the TPUVM.
